src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseStringEqualsOverStringCompare.cs

```diff
@@ -0,0 +1,248 @@
+    /// </list>
+    /// See all the <c>Is...Case</c> methods to see the <c>string.Compare</c> overloads that are supported.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class UseStringEqualsOverStringCompare : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2251";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(Resx.UseStringEqualsOverStringCompareTitle), Resx.ResourceManager, typeof(Resx));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(Resx.UseStringEqualsOverStringCompareMessage), Resx.ResourceManager, typeof(Resx));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(Resx.UseStringEqualsOverStringCompareDescription), Resx.ResourceManager, typeof(Resx));
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Usage,
+            RuleLevel.IdeHidden_BulkConfigurable,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out var symbols))
+                return;
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.Binary);
+            return;
+
+            //  Local functions
+
+            void AnalyzeOperation(OperationAnalysisContext context)
+            {
+                var operation = (IBinaryOperation)context.Operation;
+                foreach (var selector in CaseSelectors)
+                {
+                    if (selector(operation, symbols))
+                    {
+                        context.ReportDiagnostic(operation.CreateDiagnostic(Rule));
+                        return;
+                    }
+                }
+            }
+        }
+
+        internal sealed class RequiredSymbols
+        {
+            //  Named-constructor 'TryGetSymbols' inits all properties or doesn't construct an instance.
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+            private RequiredSymbols() { }
```

Comment:
This is a clarity/style concern for me. It's likely easier to just have a constructor with paramters needed than to rely on all properties being init only. Or convert this to a record?

Reply:
Agreed. I'll change it.